File: 499-The Maze III.py
from heapq import heappush, heappop
class Solution(object):
    def findShortestWay(self, maze, ball, hole):
        """
        :type maze: List[List[int]]
        :type ball: List[int]
        :type hole: List[int]
        :rtype: str
        """
        
        row, col = len(maze), len(maze[0])
        dist = [[float('inf')]*col for _ in range(row)]
        pq = [(0, "", ball[0], ball[1])]
        res = ""
        
        rMap = {"d":1, "l":0, "r":0, "u":-1}
        cMap = {"d":0, "l":-1, "r":1, "u":0}
        
        def nextNode(direction, r, c):
            ans = 0
            while True:
                dr, dc = rMap[direction], cMap[direction]
                if r+dr>=0 and r+dr<row and c+dc>=0 and c+dc<col and maze[r+dr][c+dc] == 0:
                    r = r + dr
                    c = c + dc
                    ans += 1
                    if r==hole[0] and c==hole[1]:
                        return ans, r, c
                else:
                    break
            return ans, r, c
        
        while pq:
            d, string, r, c = heappop(pq)
            if d > dist[r][c]:
                continue
            else:
                dist[r][c] = d
            
            if r == hole[0] and c == hole[1]:
                res = string
                break
            
            for direction in "dlru":
                toMove, nr, nc = nextNode(direction, r, c)
                if d + toMove >= dist[nr][nc]:
                    continue
                heappush(pq, (d+toMove, string+direction, nr, nc))
        
        if not res:
            return "impossible"
        
        return res

File: C++/499-The Maze III.cpp
// O(mn*log(mn)), note that heap size is O(mn)
class Solution {
public:
    unordered_map<char, pair<int, int>> directions = {{'d',{1, 0}}, {'l',{0, -1}}, {'r',{0, 1}}, {'u',{-1, 0}}};
    string findShortestWay(vector<vector<int>>& maze, vector<int>& ball, vector<int>& hole) {
        int row = maze.size(), col = maze[0].size();
        priority_queue<tuple<int, string, int ,int, int, int>, vector<tuple<int, string, int ,int, int, int>>, greater<>> pq;

        pq.push({0, "", ball[0], ball[1], -1, -1});
        vector<vector<bool>> visited(row, vector<bool>(col, false));

        bool found = false;
        while (!pq.empty()) {
            auto [dist, path, r, c, prevR, prevC] = pq.top();
            pq.pop();

            if (visited[r][c]) continue;
            visited[r][c] = true;

            if (r == hole[0] && c == hole[1]) {
                return path;
            }

            for (auto [direction, movePair]:directions) {
                int nxtR = r, nxtC = c;
                int rMove = movePair.first, cMove = movePair.second;

                while (nxtR >= 0 && nxtR < row && nxtC >= 0 && nxtC < col && !(nxtR == hole[0] && nxtC == hole[1]) && maze[nxtR][nxtC] == 0) {
                    nxtR += rMove;
                    nxtC += cMove;
                }

                if (!(nxtR == hole[0] && nxtC == hole[1])) {
                    nxtR -= rMove;
                    nxtC -= cMove;
                }
                
                if (visited[nxtR][nxtC]) continue;
                int moved = abs(nxtR-r) + abs(nxtC-c);
                string newPath = path + direction;
                pq.push({dist+moved, newPath, nxtR, nxtC, r, c});
            }
        }

        return "impossible";
    }
};